### PR TITLE
Create coda file

### PR DESCRIPTION
**Pull request by princelySid**
Takes a TracedData Object and converts it to CODA format and outputs that as a CSV

### CODE REVIEW

**Comment by as2388** [comment]
@princelySid please can you merge master into this branch. Then I can review it without having to re-review your code from the previous pull request.

**Comment by princelySid** [comment]
@as2388 I merged it with concatenate_flows before I merged it with master, it should be up to date

**Comment by as2388** [comment]
@princelySid I'm still seeing e.g. concatenate_pdms in the diff

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (as2388) [inline .vscode/settings.json]: check with luke that it's ok to commit ide configurations to the git repo

File: concatenate_flows/docker-run.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+set -e
+
+IMAGE_NAME=maap-concatenate_pdms
```
Comment (as2388) [inline concatenate_flows/docker-run.sh]: prefer hyphens in image names

File: concatenate_flows/docker-run.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+# Check that the correct number of arguments were provided.
+if [ $# -ne 76 ]; then
```
Comment (as2388) [inline concatenate_flows/docker-run.sh]: How did you manage to test this code without this line failing?

File: concatenate_flows/docker-run.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    echo "Usage: sh docker-run.sh <user> <pdm1> <pdm2> <pdm3> <pdm4> <pdm5> <combined_json>"
+    exit
+fi
+
+# Assign the program arguments to bash variables.
+USER = $1
+PDM1 = $2
```
Comment (as2388) [inline concatenate_flows/docker-run.sh]: prefer file args to start with INPUT_ or OUTPUT_ so it's clear which is which

File: create_coda_files/Pipfile [hunk 5]
```diff
@@ -0,0 +1,13 @@
+[[source]]
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+coredatamodules = {editable = true, ref = "v0.7.0", git = "https://www.github.com/AfricasVoices/CoreDataModules"}
```
Comment (as2388) [inline create_coda_files/Pipfile]: Fine with this being 0.7.0 for now, but you should to check the release notes for 0.7.1 and 0.7.2 and consider upgrading before you interact with coda, particularly when importing back from coded coda files.

File: create_coda_files/create_coda_files.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        prev_coda_path = args.prev_coda_path
+    else:
+        prev_coda_path = None
+    
+    # Load data from JSON file
+    with open(json_input_path, "r") as f:
+        show_messages = TracedDataJsonIO.import_json_to_traced_data_iterable(f)
+
+    # Filter out test messages sent by AVF.
+    show_messages = [td for td in show_messages if not td.get("test_run", False)]
```
Comment (as2388) [inline create_coda_files/create_coda_files.py]: Don't you want this to be commented out (with a TODO here and an issue against the repo) until you finish testing?

File: create_coda_files/docker-run.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+
+set -e
+
+IMAGE_NAME=maap-get_coda_files
+
+# Check that the correct number of arguments were provided.
+if [[ $# -lt 5 || $# -gt 6 ]]; then
+    echo "Usage: sh docker-run.sh <user> <json-input-path> <flow-name> <variable-name> <coda-output-path> <--prev-coda-input-path>"
```
Comment (as2388) [inline create_coda_files/docker-run.sh]: Optional arguments are conventionally come first.

File: create_coda_files/docker-run.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+
+set -e
+
+IMAGE_NAME=maap-get_coda_files
```
Comment (as2388) [inline create_coda_files/docker-run.sh]: hyphens

File: create_coda_files/docker-run.sh [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    echo "Usage: sh docker-run.sh <user> <json-input-path> <flow-name> <variable-name> <coda-output-path> <--prev-coda-input-path>"
+    exit
+fi
+
+# Assign the program arguments to bash variables.
+USER=$1
+INPUT_JSON=$2
+FLOW_NAME=$3
+VARIABLE_NAME=$4
+OUTPUT_CODA=$5
```
Comment (as2388) [inline create_coda_files/docker-run.sh]: You need to assign PREV_CODA and pass it into the container too

File: scripts/test_contacts.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+[
+    "202baa75-319f-4b9e-9709-fb9d011c4f87",
+    "906fa315-9722-4391-86e1-07279ea6ff91",
+    "6670e759-1a58-4bbf-a5cf-b91bdea5a9e0",
+    "d03e6248-fb2a-4393-9a57-4fb949baecf4"
+]
+
```
Comment (as2388) [inline scripts/test_contacts.json]: no new line at end of file

File: create_coda_files/create_coda_files.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            {eat_key: eat_time},
+            Metadata(user, Metadata.get_call_location(), time.time())
+        )
+
+    # Output messages to Coda
+    IOUtils.ensure_dirs_exist_for_file(coda_output_path)
+    if prev_coda_path:
+        # TODO: Modifying this line once the coding frame has been developed to include lots of Nones feels a bit
+        # TODO: cumbersome. We could instead modify export_traced_data_iterable_to_coda to support a prev_f argument.
+        scheme_keys = {"Relevance": None, "Code 1": None, "Code 2": None, "Code 3": None, "Code 4": None}
```
Comment (as2388) [inline create_coda_files/create_coda_files.py]: This needs to be set to the scheme used when known. For now, I would set this to empty, with a TODO and an issue against the repo saying that this needs to be updated.

File: .vscode/settings.json (hunk 1 above)
Comment (princelySid) [reply .vscode/settings.json]: This was missed by my .gitignore changing it.

File: concatenate_flows/docker-run.sh (hunk 3 above)
Comment (princelySid) [reply concatenate_flows/docker-run.sh]: It didn't, hadn't tested it yet

File: create_coda_files/Pipfile (hunk 5 above)
Comment (princelySid) [reply create_coda_files/Pipfile]: Noted, upgrading

File: concatenate_flows/docker-run.sh (hunk 3 above)
Comment (princelySid) [reply concatenate_flows/docker-run.sh]: Now tested

File: create_coda_files/docker-run.sh (hunk 7 above)
Comment (princelySid) [reply create_coda_files/docker-run.sh]: Is this in the python file as well or just in the bash script?

Comment (princelySid) [reply create_coda_files/docker-run.sh]: And how does it affect how the script runs?

Comment (as2388) [reply create_coda_files/docker-run.sh]: Both.

`python script.py --optional x a b c` instead of `python script.py a b c --optional x`

Comment (as2388) [reply create_coda_files/docker-run.sh]: Optionals are a little painful in bash - here's an example of how to do it https://github.com/AfricasVoices/RapidProTools/blob/master/fetch_contacts/docker-run.sh

Comment (as2388) [reply create_coda_files/docker-run.sh]: Alternatively, you can make the prev-coda-input-path compulsory, but allow that directory to be empty, which is what REACH did.

Comment (princelySid) [reply create_coda_files/docker-run.sh]: Lemme do that, looks easier.

File: .vscode/settings.json (hunk 1 above)
Comment (as2388) [reply .vscode/settings.json]: This file is still here

File: create_coda_files/create_coda_files.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                        help="Path to a Coda file containing previously coded messages")
+
+    args = parser.parse_args()
+    user = args.user
+    json_input_path = args.json_input_path
+    flow_name = args.flow_name
+    variable_name = args.variable_name
+    coda_output_path = args.coda_output_path
+    prev_coda_path = args.prev_coda_path
+    if os.path.exists(prev_coda_path):
```
Comment (as2388) [inline create_coda_files/create_coda_files.py]: If you did `if not` here you wouldn't need the weird empty if branch.

File: scripts/test_contacts.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+[    
+    "202baa75-319f-4b9e-9709-fb9d011c4f87",
+    "906fa315-9722-4391-86e1-07279ea6ff91",
+    "6670e759-1a58-4bbf-a5cf-b91bdea5a9e0",
+    "d03e6248-fb2a-4393-9a57-4fb949baecf4"
+]
```
Comment (as2388) [inline scripts/test_contacts.json]: no new line at end of file

File: create_coda_files/create_coda_files.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    # Filter out test messages sent by AVF.
+    show_messages = [td for td in show_messages if not td.get("test_run", False)]
+
+    # Filter for runs which contain this key.
+    show_message_key = "{} (Text) - {}".format(variable_name, flow_name)
+    show_messages = [td for td in show_messages if show_message_key in td]
+
+    # Convert date/time of messages to EAT
+    utc_key = "{} (Time) - {}".format(variable_name, flow_name)
+    eat_key = "{} (Time EAT) - {}".format(variable_name, flow_name)
```
Comment (as2388) [inline create_coda_files/create_coda_files.py]: I thought this was going to be moved to be near the interface exporter?

**Pull request merged**